lib/pos_snapshot_entries.go

```diff
@@ -0,0 +1,764 @@
+			utxoViewValidatorEntries = append(utxoViewValidatorEntries, validatorEntry)
+		}
+	}
+	// Pull top N active ValidatorEntries from the database (not present in the UtxoView).
+	// Note that we will skip validators that are present in the view because we pass
+	// utxoViewValidatorEntries to the function.
+	dbValidatorEntries, err := DBGetSnapshotTopActiveValidatorsByStake(
+		bav.Handle, bav.Snapshot, limit, snapshotAtEpochNumber, utxoViewValidatorEntries,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(err, "GetSnapshotTopActiveValidatorsByStake: error retrieving entries from db: ")
+	}
+	// Cache top N active ValidatorEntries from the db in the UtxoView.
+	for _, validatorEntry := range dbValidatorEntries {
+		// We only pull ValidatorEntries from the db that are not present in the
+		// UtxoView. As a sanity check, we double-check that the ValidatorEntry
+		// is not already in the UtxoView here.
+		mapKey := SnapshotValidatorMapKey{
+			SnapshotAtEpochNumber: snapshotAtEpochNumber, ValidatorPKID: *validatorEntry.ValidatorPKID,
+		}
+		if _, exists := bav.SnapshotValidatorEntries[mapKey]; !exists {
+			bav._setSnapshotValidatorEntry(validatorEntry, snapshotAtEpochNumber)
+		}
+	}
+	// Pull !isDeleted, active ValidatorEntries from the UtxoView with stake > 0.
+	var validatorEntries []*ValidatorEntry
+	for mapKey, validatorEntry := range bav.SnapshotValidatorEntries {
+		if mapKey.SnapshotAtEpochNumber == snapshotAtEpochNumber &&
+			!validatorEntry.isDeleted &&
+			validatorEntry.Status() == ValidatorStatusActive &&
+			!validatorEntry.TotalStakeAmountNanos.IsZero() {
+			validatorEntries = append(validatorEntries, validatorEntry)
+		}
+	}
+	// Sort the ValidatorEntries DESC by TotalStakeAmountNanos.
+	sort.SliceStable(validatorEntries, func(ii, jj int) bool {
+		stakeCmp := validatorEntries[ii].TotalStakeAmountNanos.Cmp(validatorEntries[jj].TotalStakeAmountNanos)
+		if stakeCmp == 0 {
+			// Use ValidatorPKID as a tie-breaker if equal TotalStakeAmountNanos.
+			return bytes.Compare(
+				validatorEntries[ii].ValidatorPKID.ToBytes(),
+				validatorEntries[jj].ValidatorPKID.ToBytes(),
+			) > 0
+		}
+		return stakeCmp > 0
+	})
+	// Return top N.
+	upperBound := int(math.Min(float64(limit), float64(len(validatorEntries))))
+	return validatorEntries[0:upperBound], nil
+}
+
+func (bav *UtxoView) _setSnapshotValidatorEntry(validatorEntry *ValidatorEntry, snapshotAtEpochNumber uint64) {
+	if validatorEntry == nil {
+		glog.Errorf("_setSnapshotValidatorEntry: called with nil entry, this should never happen")
+		return
+	}
+	mapKey := SnapshotValidatorMapKey{
+		SnapshotAtEpochNumber: snapshotAtEpochNumber, ValidatorPKID: *validatorEntry.ValidatorPKID,
+	}
+	bav.SnapshotValidatorEntries[mapKey] = validatorEntry.Copy()
+}
+
+func (bav *UtxoView) _flushSnapshotValidatorEntriesToDbWithTxn(txn *badger.Txn, blockHeight uint64) error {
+	for mapKey, validatorEntry := range bav.SnapshotValidatorEntries {
```

Comment:
I think there is a bug here. You need to first DELETE all the entries then PUT all the entries. I think the code as it's written now will have messed up indexes after a few flushes. Consider the following example:
* PUT (epoch=1, validatorpkid=validator1, stakeAmount=5)
  - Creates 5->validator1
* UPDATE (epoch=1, validatorpkid=validator1, stakeAmount=6)
  - Creates 6->validator1 WITHOUT deleting 5->validator1

In order to hit this bug, you'd need to do two snapshot flushes within the same epoch which I don't *think* can happen. But adding the delete makes it bullet-proof and additionally makes it conform to all our other flushes. I would do this for all of them.

Also when you add the DELETE, I'd do it by (epoch, pkid) so that you just look up all the entries that correspond to that validator in that snapshot and kill them before you do a PUT.

Reply:
> In order to hit this bug, you'd need to do two snapshot flushes within the same epoch which I don't think can happen. 
That's right. That should never happen. Snapshot values should be immutable once they're set in the OnEpochCompleteHook. For that reason, I didn't add delete functionality. IMO it's wasted db ops. But I will add for consistency. Easy enough to rip out later.

Reply:
I'm still on the fence about this pattern. A snapshot validator shouldn't have their TotalStakeAmountNanos changed for example.

Reply:
A snapshot validator's TotalStakeAmountNanos should never change. But I did add this delete then set pattern.